File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/virology/CtaToken.java
package uk.nhs.nhsx.virology;

import uk.nhs.nhsx.core.ValueType;
import uk.nhs.nhsx.core.random.crockford.CrockfordDammRandomStringGenerator.DammChecksum;

import static uk.nhs.nhsx.core.random.crockford.CrockfordDammRandomStringGenerator.checksum;

public class CtaToken extends ValueType<CtaToken> {

    private static final DammChecksum checksum = checksum();

    private CtaToken(String value) {
        super(value);
        if (!checksum.validate(value)) throw new IllegalArgumentException("validation error: Token Failed Validation");
    }

    public static CtaToken of(String value) {
        return new CtaToken(value);
    }
}

File: src/aws/lambdas/incremental_distribution/infra/src/main/kotlin/uk/nhs/nhsx/domain/CtaToken.kt
package uk.nhs.nhsx.domain

import dev.forkhandles.values.StringValue
import dev.forkhandles.values.StringValueFactory
import uk.nhs.nhsx.core.random.crockford.CrockfordDammRandomStringGenerator

class CtaToken private constructor(value: String) : StringValue(value) {
    companion object : StringValueFactory<CtaToken>(
        ::CtaToken,
        validation = CrockfordDammRandomStringGenerator.checksum()::validate
    )
}
